### PR TITLE
Shadowban plugin

### LINKED ISSUE #222: server wide shadow ban
ability to ban a user from all non class channels

### PR DESCRIPTION
**Pull request by tanndlin**
fixes #222
Removes user ability to see all general text channels, and voice channels
They can still see class channels, and the `UCF CS, ECE, AND IT` category

### CODE REVIEW

File: src/app/plugins/shadowban.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  public validate(message: IMessage, args: string[]) {
+    return args && args.length > 1;
+  }
+
+  public async execute(message: IMessage, args?: string[]) {
+    if (!args) {
+      return;
+    }
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:34]: should be able to skip this by making the param not `?` and `validate` is always called before execute

File: src/app/plugins/shadowban.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    const [subCommand, ...targetUser] = args;
+    const member = this.container.guildService
+      .get()
+      .members.filter((m) => m.user.tag === targetUser.join(' '))
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:39]: dont call the join() N times

File: src/app/plugins/shadowban.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    return args && args.length > 1;
+  }
+
+  public async execute(message: IMessage, args?: string[]) {
+    if (!args) {
+      return;
+    }
+
+    const [subCommand, ...targetUser] = args;
+    const member = this.container.guildService
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:37]: feel free to use the `resolveUser` from Moderation Service.

File: src/app/plugins/shadowban.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    if (subCommand === 'ban') {
+      this._banUser(message, member.user);
+      return;
+    } else if (subCommand === 'unban') {
+      this._unbanUser(message, member.user);
+      return;
+    } else {
+      message.reply('Invalid subcommand\nTry: `!shadowban <ban|unban> <user>`');
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:54]:
```
${this.description}
```
instead of hardcode

File: src/app/plugins/shadowban.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    const catsToBan: CategoryChannel[] = [];
+    const categories: Collection<
+      string,
+      CategoryChannel
+    > = this.container.guildService
+      .get()
+      .channels.filter((chan) => chan.type === 'category') as Collection<string, CategoryChannel>;
+
+    categories.forEach((cat) => {
+      if (!this.BANNED_CATEGORIES.some((n) => cat.name.toUpperCase() === n)) {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:86]: only UpperCase once

File: src/app/plugins/shadowban.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  private _getChannelsToBan(): CategoryChannel[] {
+    const catsToBan: CategoryChannel[] = [];
+    const categories: Collection<
+      string,
+      CategoryChannel
+    > = this.container.guildService
+      .get()
+      .channels.filter((chan) => chan.type === 'category') as Collection<string, CategoryChannel>;
+
+    categories.forEach((cat) => {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:85]: reduce()

File: src/app/plugins/shadowban.plugin.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+  private _unbanUser(message: IMessage, user: User) {
+    this._getChannelsToBan().forEach((chan) => {
+      chan.permissionOverwrites.get(user.id)?.delete();
+    });
+
+    message.reply(`${user.tag} has been unshadowbanned`);
+  }
+
+  private _getChannelsToBan(): CategoryChannel[] {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:76]: you could make this method take a parameter that is a function F(Channel) -> void and then at the end use reduce to apply that function... then in ban and unban you dont need the forEach explicitly.

**Review by c650 — COMMENTED** [review]
does this work if a channel within a category doesnt have perms synced?

**Comment by tanndlin** [comment]
No, I was going to do a channel by channel approach but figured all the channels were synced and it would be neater in code this way.

**Comment by tanndlin** [comment]
When I get back home I'll look into it

File: src/app/plugins/shadowban.plugin.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+      this._unbanUser(message, member.user);
+      return;
+    } else {
+      message.reply('Invalid subcommand\nTry: `!shadowban <ban|unban> <user>`');
+    }
+  }
+
+  private _banUser(message: IMessage, user: User) {
+    this._getChannelsToBan().forEach((chan) => {
+      chan.overwritePermissions(user, {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:60]: if this is async we should do a map and then await Promise.all

**Comment by tanndlin** [comment]
Doing a callback was the weirdest thing I've ever done, but I kinda like it 🤔

File: src/app/plugins/shadowban.plugin.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      .get()
+      .channels.filter((chan) => chan.type === 'category') as Collection<string, CategoryChannel>;
+
+    const catsToBan = categories.filter((cat: CategoryChannel) => {
+      const chanName = cat.name.toUpperCase();
+      return this.BANNED_CATEGORIES.some((n) => chanName === n);
+    });
+
+    catsToBan
+      .reduce(function(acc: GuildChannel[], cat: CategoryChannel) {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:68]: prefer `() =>` over `function ()`

File: src/app/plugins/shadowban.plugin.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        return [...acc, ...cat.children.array()];
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:69]: why not 

```
acc.push(...cat.children.array())
return acc
```

File: src/app/plugins/shadowban.plugin.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+      }, [])
+      .forEach((chan) => callback(chan, user));
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:71]: might as well do this part in the reduce

File: src/app/plugins/shadowban.plugin.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  private _banUser(chan: GuildChannel, user: User) {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:74]: for extra coolness:

```
private _banUser(user: User) { 
  return (chan: GuildChannel) => {
    chan.overwritePermissions(user, ...);
  };
}
```

then you can pass `_banUser(user)` and then the callback is `callback(chan)`...

or better yet! 
```
chans.forEach(callback)
```

File: src/app/plugins/shadowban.plugin.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    chan.overwritePermissions(user, {
+      VIEW_CHANNEL: false,
+    });
+  }
+
+  private _unbanUser(chan: GuildChannel, user: User) {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:80]: saa

File: src/app/plugins/shadowban.plugin.ts (hunk 11 above)
Comment (tanndlin) [reply src/app/plugins/shadowban.plugin.ts:71]: I'm struggling to see how to do it in the reduce,
`acc,push(...cat.children.array().forEach(...)`?

Comment (c650) [reply src/app/plugins/shadowban.plugin.ts:71]: something like that but change forEach to `map` and then you can reduce the categories to the results of the channel overrides. I thought that channel overrides were async so they'd have to be awaited?

then you can do

```js
promises = ...reduce();
await Promise.all(promises);

```

File: src/app/plugins/shadowban.plugin.ts [hunk 14]
```diff
@@ -64,20 +71,25 @@ export class ShadowBanPlugin extends Plugin {
...
+    await Promise.all(promises);
   }
 
-  private _banUser(chan: GuildChannel, user: User) {
-    chan.overwritePermissions(user, {
-      VIEW_CHANNEL: false,
-    });
+  private _banUser(user: User) {
+    return (chan: GuildChannel) => {
+      chan.overwritePermissions(user, {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:77]: `overwritePermissions` returns a promise. You should await it.

```
return async (chan) => {
  await chan.overwritePermissions(...);
}
```

https://discord.js.org/#/docs/main/stable/class/GuildChannel?scrollTo=overwritePermissions

File: src/app/plugins/shadowban.plugin.ts [hunk 15]
```diff
@@ -64,20 +71,25 @@ export class ShadowBanPlugin extends Plugin {
...
+        VIEW_CHANNEL: false,
+      });
+    };
   }
 
-  private _unbanUser(chan: GuildChannel, user: User) {
-    chan.permissionOverwrites.get(user.id)?.delete();
+  private _unbanUser(user: User) {
+    return (chan: GuildChannel) => {
```
Comment (c650) [inline src/app/plugins/shadowban.plugin.ts:84]: saa

File: src/app/plugins/shadowban.plugin.ts (hunk 14 above)
Comment (tanndlin) [reply src/app/plugins/shadowban.plugin.ts:77]: Still slightly figuring out promises lol, thought that `await Promise.all` was enough

**Review by c650 — COMMENTED** [review]
now await the `applyToChannels` in `execute` and you're good to go.

**Review by c650 — APPROVED** [review]

**Pull request merged**